### PR TITLE
Refactor the gagdets folder 

### LINKED ISSUE #124: Refactor DUSK contract gadgets
Following an in person review with @toghrulmaharramov and a telephone consultation with @ZER0, it has been decided that something on the DUSK contract circuits ought to be changed. 

The first required change is the hashing method for the preimage circuit. Currently, the preimage gadget hashes the value commitment, position and stealth address within the note. This will be replaced by hashing more fields from the note, these will be the encrypted data, the R within the stealth address, the nonce and the note type.

The second needed feature is a schnorr gadget, which allows the user to prove the DLP of a signature given a public key pair.

### PR DESCRIPTION
**Pull request by Bounce23**
As per the refactor required in the dusk contract, this PR will address the gadgets which permeate the circuits.

The first of these is the inner workings of the preimage gadget, which cannot be tied to the note in the current specs. This PR will refactor the the preimage gadget and its uses within rust.
The second is a schnorr signature gadget, which allows us to prove the knowledge of DLP given a signature.

This closes #124.

### CODE REVIEW

File: contracts/transfer/circuits/src/gadgets/preimage.rs [hunk 1]
```diff
@@ -7,26 +7,39 @@
 use dusk_plonk::prelude::*;
 use plonk_gadgets::AllocatedScalar;
 use poseidon252::sponge::sponge::sponge_hash_gadget;
+use std::convert::TryInto;
 
 /// Prove knowledge of the preimage of a note,
 /// used as input for a transaction.
 #[allow(non_snake_case)]
```
Comment (CPerezz) [inline contracts/transfer/circuits/src/gadgets/preimage.rs:14]: Why is this needed?  I don't see any uppercase variables.

File: contracts/transfer/circuits/src/gadgets/preimage.rs [hunk 2]
```diff
@@ -7,26 +7,39 @@
...
 pub fn input_preimage(
     composer: &mut StandardComposer,
+    note_type: AllocatedScalar,
     value_commitment_x: AllocatedScalar,
     value_commitment_y: AllocatedScalar,
-    pos: AllocatedScalar,
     pk_r_x: AllocatedScalar,
     pk_r_y: AllocatedScalar,
+    randomness_x: AllocatedScalar,
+    randomness_y: AllocatedScalar,
```
Comment (CPerezz) [inline contracts/transfer/circuits/src/gadgets/preimage.rs:23]: Why don't you expect a `plonk::ecc::Point` instead of this. Will make your code much easier and you have access to it's coordinates so that you can re-use them without breaking any linking.

It would be something like:
```rust
use dusk_plonk::constraint_system::ecc::Point as PlonkPoint;
pub fn input_preimage(
     composer: &mut StandardComposer;
     note_type: AllocatedScalar,
     value_commitment: PlonkPoint,
     pkr: PlonkPoint,
     randomness: PlonkPoint,
     // Whatever else....
) { 
     // Use the coordinates of the `PlonkPoint` which is a `JubJubAffine` 
     // kinda off "allocated" and so you don't break any linking.
}

File: contracts/transfer/circuits/src/gadgets/schnorr.rs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        &[*R.x(), *R.y(), *R_prime.x(), *R_prime.y(), h],
+    );
+
+    let sig_1 = scalar_mul(composer, signature.var, GENERATOR_EXTENDED);
+    let sig_2 = scalar_mul(composer, signature.var, GENERATOR_NUMS_EXTENDED);
+
+    let pub_1 = scalar_mul(composer, challenge, ExtendedPoint::from(pk));
+    let pub_2 = scalar_mul(composer, challenge, ExtendedPoint::from(pk_prime));
+
+
```
Comment (CPerezz) [inline contracts/transfer/circuits/src/gadgets/schnorr.rs:40]: I guess that here you still need to check whether the points you got from the scalar mul are equal right??

File: contracts/transfer/circuits/src/gadgets/preimage.rs (hunk 1 above)
Comment (CPerezz) [reply contracts/transfer/circuits/src/gadgets/preimage.rs:17]: Still not answered

File: contracts/transfer/circuits/src/gadgets/preimage.rs [hunk 4]
```diff
@@ -4,37 +4,52 @@
...
+    value_commitment: PlonkPoint,
+    nonce: AllocatedScalar,
+    pk_r: PlonkPoint,
+    randomness: PlonkPoint,
+    position: AllocatedScalar,
+    cipher_one: AllocatedScalar,
+    cipher_two: AllocatedScalar,
+    cipher_three: AllocatedScalar,
 ) -> Variable {
     let output = sponge_hash_gadget(
```
Comment (CPerezz) [inline contracts/transfer/circuits/src/gadgets/preimage.rs:30]: No need to assingn the result to a variable and return it. Just return the result of the function without allocating it.

File: contracts/transfer/circuits/src/gadgets/preimage.rs [hunk 5]
```diff
@@ -45,44 +60,83 @@ mod commitment_tests {
...
+            blinder,
+        );
+        let note_bytes = note.to_bytes();
+        let cipher = &note_bytes[note_bytes.len() - 96..];
+        let arr1: [u8; 32] = cipher[..32].try_into().expect("Invalid length");
+        let cipher1 = BlsScalar::from_bytes(&arr1).unwrap();
+        let arr2: [u8; 32] = cipher[32..64].try_into().expect("Invalid length");
+        let cipher2 = BlsScalar::from_bytes(&arr2).unwrap();
+        let arr3: [u8; 32] = cipher[64..].try_into().expect("Invalid length");
+        let cipher3 = BlsScalar::from_bytes(&arr3).unwrap();
```
Comment (CPerezz) [inline contracts/transfer/circuits/src/gadgets/preimage.rs:87]: Is this needed anymore?

File: contracts/transfer/circuits/src/gadgets/preimage.rs [hunk 6]
```diff
@@ -45,44 +60,83 @@ mod commitment_tests {
...
+        let nonce = JubJubScalar::from(345 as u64);
+        let note = Note::deterministic(
+            NoteType::Obfuscated,
+            &r,
+            nonce,
+            &psk,
+            value,
+            blinder,
+        );
+        let note_bytes = note.to_bytes();
```
Comment (CPerezz) [inline contracts/transfer/circuits/src/gadgets/preimage.rs:80]: Is this needed?

File: contracts/transfer/circuits/src/gadgets/preimage.rs [hunk 7]
```diff
@@ -45,44 +60,83 @@ mod commitment_tests {
...
+        let cipher3 = BlsScalar::from_bytes(&arr3).unwrap();
+
         let note_hash = note.hash();
 
         // Generate Composer & Public Parameters
         let pub_params =
-            PublicParameters::setup(1 << 15, &mut rand::thread_rng())?;
-        let (ck, vk) = pub_params.trim(1 << 14)?;
+            PublicParameters::setup(1 << 17, &mut rand::thread_rng())?;
+        let (ck, vk) = pub_params.trim(1 << 16)?;
```
Comment (CPerezz) [inline contracts/transfer/circuits/src/gadgets/preimage.rs:94]: Does it need to be that big the parameters used?

File: contracts/transfer/circuits/src/gadgets/schnorr.rs [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+use dusk_plonk::prelude::*;
+use plonk_gadgets::AllocatedScalar;
+use poseidon252::sponge::sponge::sponge_hash_gadget;
+use std::convert::TryInto;
+
+/// Utilises a Schnorr signature scheme,
+/// to prove the knowledge of a discrete
+/// log for a given public key.
+#[allow(non_snake_case)]
+pub fn schnorr_one_key(
```
Comment (CPerezz) [inline contracts/transfer/circuits/src/gadgets/schnorr.rs:22]: I'd add _gadget to the fn name so that is clear that is a gadget.

File: contracts/transfer/circuits/src/gadgets/schnorr.rs [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+    composer.assert_equal_point(add, R);
+}
+
+/// Utilises a Schnorr signature scheme,
+/// to prove the knowledge of the discrete
+/// log for given keys in a public key pair.
+/// Also verifying that both keys share the
+/// same discrete log.
+#[allow(non_snake_case)]
+pub fn schnorr_two_keys(
```
Comment (CPerezz) [inline contracts/transfer/circuits/src/gadgets/schnorr.rs:50]: Same as above.

File: contracts/transfer/circuits/src/gadgets/schnorr.rs [hunk 10]
```diff
@@ -0,0 +1,206 @@
...
+pub fn schnorr_one_key(
+    composer: &mut StandardComposer,
+    signature: AllocatedScalar,
+    R: PlonkPoint,
+    pk: PlonkPoint,
+    message: AllocatedScalar,
+) {
+    let h = sponge_hash_gadget(composer, &[message.var]);
+    let c = sponge_hash_gadget(composer, &[*R.x(), *R.y(), h]);
+    let b = BlsScalar::zero();
```
Comment (CPerezz) [inline contracts/transfer/circuits/src/gadgets/schnorr.rs:31]: This can be avoided and unse BlsScalar::zero below.

File: contracts/transfer/circuits/src/gadgets/schnorr.rs [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+    let b = composer.add_witness_to_circuit_description(b);
+
+    let challenge = composer.xor_gate(c, b, 250);
+
+    let sig =
+        fixed_base_scalar_mul(composer, signature.var, GENERATOR_EXTENDED);
+    let p = variable_base_scalar_mul(composer, challenge, pk);
+
+    let add = sig.point().add(composer, *p.point());
+    composer.assert_equal_point(add, R);
```
Comment (CPerezz) [inline contracts/transfer/circuits/src/gadgets/schnorr.rs:41]: If this is a public point, the constraint should be added outside.

File: contracts/transfer/circuits/src/gadgets/schnorr.rs [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+        fixed_base_scalar_mul(composer, signature.var, GENERATOR_EXTENDED);
+    let sig_2 =
+        fixed_base_scalar_mul(composer, signature.var, GENERATOR_NUMS_EXTENDED);
+    let pub_1 = variable_base_scalar_mul(composer, challenge, pk);
+    let pub_2 = variable_base_scalar_mul(composer, challenge, pk_prime);
+
+    let add_1 = sig_1.point().add(composer, *pub_1.point());
+    composer.assert_equal_point(add_1, R);
+    let add_2 = sig_2.point().add(composer, *pub_2.point());
+    composer.assert_equal_point(add_2, R_prime);
```
Comment (CPerezz) [inline contracts/transfer/circuits/src/gadgets/schnorr.rs:78]: Same as above. If the points are public, then we need to add the checks outside.

**Review by CPerezz — CHANGES_REQUESTED** [review]
Some nits/questions

File: contracts/transfer/circuits/src/gadgets/schnorr.rs (hunk 11 above)
Comment (Bounce23) [reply contracts/transfer/circuits/src/gadgets/schnorr.rs:41]: Nothing is public According to Tog.

File: contracts/transfer/circuits/src/gadgets/schnorr.rs (hunk 12 above)
Comment (Bounce23) [reply contracts/transfer/circuits/src/gadgets/schnorr.rs:78]: Nothing is public according to Tog.

**Review by CPerezz — APPROVED** [review]
LGTM. Please remove the unnecessary `#[allow(non_snake_case)]` and wait for the CI pr to be merged so that you can pass the CI and merge this.
